EventPhases/explainer.md

```diff
@@ -0,0 +1,501 @@
+# Opt-In Event Phases For Reliably Fast DOM Operations
+
+## Authors:
+
+- [Alex Russell](https://infrequently.org/about-me/) &lt;alexrussell@microsoft.com&gt;, &lt;slightlyoff@chromium.org&gt;
+
+## Participate
+
+- [Issue tracker](https://github.com/MicrosoftEdge/MSEdgeExplainers/issues/new?assignees=slightlyoff&labels=EventPhases&title=%5BEventPhases%5D+%3CTITLE+HERE%3E)
+
+<!-- - [Discussion forum] -->
+
+## Introduction
+
+["Layout thrashing"](https://web.dev/articles/avoid-large-complex-layouts-and-layout-thrashing) is a widely acknowledged performance problem in web application development. It has been documented [across](https://webperf.tips/tip/layout-thrashing/) [many](https://www.afasterweb.com/2015/10/05/how-to-thrash-your-layout/) [years](https://kellegous.com/j/2013/01/26/layout-performance/), generating both comprehensive [documentation on how to manually avoid it](https://gist.github.com/paulirish/5d52fb081b3570c81e3a) and [tooling to highlight when it occurs.](https://developer.chrome.com/docs/devtools/performance#find_the_bottleneck)
+
+This problem is also the proximate cause of the design of many of today's functional-reactive frameworks (although, in practice, they do not guard against it effectively).
+
+In the worst cases, layout thrashing is a key contributor to slow interactions (diagnosed most easily via the [INP metric](https://web.dev/articles/inp)). Innocent-looking code that interleaves style reads and DOM mutations can trigger layout more than once per painted frame.
+
+### How Does This Happen?
+
+[Joe Liccini's excellent blog post](https://webperf.tips/tip/layout-thrashing/) covers the underlying mechanics in detail, but to recap, the phases of generating a frame from changes to CSS and DOM (per [Paul Lewis' classic article](https://web.dev/articles/rendering-performance)) are:
+
+![The Pixel Pipeline](https://web.dev/static/articles/rendering-performance/image/the-full-pixel-pipeline-45b24543207ea_1440.jpg)
```

Comment:
How do we feel about embedding external images in this repo? Probably copy locally?

Reply:
Good call. Fixing.

Reply:
Fixed.